selfdrive/boardd/boardd.cc

```diff
@@ -148,175 +112,94 @@ void *safety_setter_thread(void *s) {
 
   capnp::FlatArrayMessageReader cmsg(amsg);
   cereal::CarParams::Reader car_params = cmsg.getRoot<cereal::CarParams>();
+  cereal::CarParams::SafetyModel safety_model = car_params.getSafetyModel();
 
-  int safety_model = int(car_params.getSafetyModel());
   auto safety_param = car_params.getSafetyParam();
-  LOGW("setting safety model: %d with param %d", safety_model, safety_param);
-
-  pthread_mutex_lock(&usb_lock);
-
-  // set in the mutex to avoid race
-  safety_setter_thread_initialized = false;
+  LOGW("setting safety model: %d with param %d", (int)safety_model, safety_param);
 
-  usb_write(dev_handle, 0xdc, safety_model, safety_param);
+  panda->set_safety_model(safety_model, safety_param);
 
-  pthread_mutex_unlock(&usb_lock);
-
-  return NULL;
+  safety_setter_thread_running = false;
+  return;
 }
 
 
-// must be called before threads or with mutex
 bool usb_connect() {
-  int err, err2;
-  unsigned char hw_query[1] = {0};
-  unsigned char fw_sig_buf[128];
-  unsigned char fw_sig_hex_buf[16];
-  unsigned char serial_buf[16];
-  const char *serial;
-  int serial_sz = 0;
-
-  ignition_last = false;
-
-  usb_close(dev_handle);
-
-  dev_handle = libusb_open_device_with_vid_pid(ctx, 0xbbaa, 0xddcc);
-  if (dev_handle == NULL) { goto fail; }
-
-  err = libusb_set_configuration(dev_handle, 1);
-  if (err != 0) { goto fail; }
-
-  err = libusb_claim_interface(dev_handle, 0);
-  if (err != 0) { goto fail; }
+  try {
+    assert(panda == NULL);
+    panda = new Panda();
+  } catch (std::exception &e) {
+    return false;
```

Comment:
seems like this can get into an infinity loop if panda isn't null for some reason?

Reply:
what do you mean? The assert won't throw an exception. It's just to verify that the previous connection was closed correctly.